File: BaekJoon/위상 정렬/2252_줄 세우기.swift
let inputs = readLine().split(separator: " ").map { Int(String($0))! }
var indegrees = [Int](repeating: 0, count: inputs[0] + 1)
var graphs = [[Int]](repeating: [], count: inputs[0] + 1)

for _ in 0..<inputs[1] {
    let infos = readLine().split(separator: " ").map { Int(String($0))! }

    graphs[infos[0]].append(infos[1])
    indegrees[infos[1]] += 1
}

topologicalSort()

func topologicalSort() {
    var queue = [Int]()

    for i in 1...inputs[0] {
        if indegrees[i] == 0 {
            queue.append(i)
        }
    }

    while !queue.isEmpty() {
        let current = queue.removeFirst()
        print(current, end = " ")

        for next in graphs[current] {
            indegrees[next] -= 1

            if indegrees[next] == 0 {
                queue.append(next)
            }
        }
    }
}

File: BaekJoon/위상 정렬/2252_줄 세우기.py
from collections import deque
import sys

input = sys.stdin.readline

def main():
    global n, m

    n, m = list(map(int, input().split()))

    graphs = [[] for _ in range(n + 1)]
    indegrees = [0 for _ in range(n + 1)]

    for _ in range(m):
        src, dst = list(map(int, input().split()))

        graphs[src].append(dst)
        indegrees[dst] += 1

    topologicalSort(graphs, indegrees)

def topologicalSort(graphs, indegrees):
    queue = deque()

    for i in range(1, n + 1):
        if indegrees[i] == 0:
            queue.append(i)

    while queue:
        current = queue.popleft()
        print(current, end =" ")

        for next in graphs[current]:
            indegrees[next] -= 1

            if indegrees[next] == 0:
                queue.append(next)

if __name__ == "__main__":
    main()